presto-main/src/main/java/com/facebook/presto/execution/scheduler/PartialResultQueryTaskTracker.java

```diff
@@ -0,0 +1,123 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.scheduler;
+
+import com.facebook.presto.execution.PartialResultQueryManager;
+import com.facebook.presto.execution.RemoteTask;
+import com.facebook.presto.execution.TaskId;
+import com.facebook.presto.execution.TaskInfo;
+import com.facebook.presto.spi.PrestoWarning;
+import com.facebook.presto.spi.WarningCollector;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static com.facebook.presto.execution.StateMachine.StateChangeListener;
+import static com.facebook.presto.spi.StandardWarningCode.PARTIAL_RESULT_WARNING;
+import static com.google.common.collect.Sets.SetView;
+import static com.google.common.collect.Sets.difference;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class PartialResultQueryTaskTracker
+        implements StateChangeListener<TaskInfo>, Comparable<PartialResultQueryTaskTracker>
+{
+    private final PartialResultQueryManager partialResultQueryManager;
+    private final double minCompletionRatioThreshold;
+    private final double timeMultiplier;
+    private final WarningCollector warningCollector;
+    private final long startTime;
+    private final Map<TaskId, RemoteTask> taskIdMap;
+    private final Set<TaskId> completedTaskIds;
+
+    private long maxEndTime;
+    private boolean taskSchedulingCompleted;
+    private boolean eligibleForPartialResults;
+
+    public PartialResultQueryTaskTracker(
+            PartialResultQueryManager partialResultQueryManager,
+            double minCompletionRatioThreshold,
+            double timeMultiplier,
+            WarningCollector warningCollector)
+    {
+        this.partialResultQueryManager = requireNonNull(partialResultQueryManager, "partialResultQueryManager is null");
+        this.minCompletionRatioThreshold = minCompletionRatioThreshold;
+        this.timeMultiplier = timeMultiplier;
+        this.warningCollector = requireNonNull(warningCollector, "warningCollector is null");
+        this.startTime = System.nanoTime();
+        this.taskIdMap = new HashMap<>();
+        this.completedTaskIds = new HashSet<>();
+    }
+
+    private double getTaskCompletionRatio()
+    {
+        return (double) completedTaskIds.size() / (double) taskIdMap.size();
+    }
+
+    public void trackTask(RemoteTask task)
+    {
+        taskIdMap.put(task.getTaskId(), task);
+    }
+
+    @Override
+    public void stateChanged(TaskInfo taskInfo)
```

Comment:
The rest of the class is called from single thread (query scheduler or partial result query manager) so they should be fine, but this method will be called concurrently so we need to be more careful.

Let's make sure:

- We use a thread-safe data structure for `completedTaskIds` (`ConcurrentHashSet`)
- Make `eligibleForPartialResults` atomic (or a latch), otherwise it is possible we would add multiple entries for the same instance, which might be okay but just to be safe. Also, this flag is mainly used to make sure we only add to query manager once, let's also change name to reflect that (e.g. `addedToQueryManager`).